shared/docs/react-native/profiling.md

```diff
@@ -0,0 +1,280 @@
+
+First I used Parashuram setup to mark the tti_complete as the latest
+message in the chat view. Then I built the app with `gradle
+installReleaseUnsigned`. This is the same as the release build, but not signed
+with our publishing key. Then I setup a simple node http server to receive the
+trace-event payload.
+
+[The simple node server](https://gist.github.com/MarcoPolo/0d3186becbce74bc8f21c9deab33dcc0)
+
+Run this a couple of times to get the starting number that you're going to try
+to improve.
+
+### Investigating slowdowns
+
+I enabled RN's systrace with the method listed above. as soon as
+I did this I saw the culprits. We load most of the app
+at startup before we can even make a single engine call. Our current routing
+model is the culprit here. Our current routing system asks us to import
+everything right from the get go.
+
+Diving slightly closer, emoji-datasource is a huge slow down. Taking about 0.3s
+itself to load. This is probably because it has to parse a lot of emoji data.
+but it's a little silly we have to do this at runtime.
+
+### Delayed Handshake
+
+I also noticed that we wait quite a bit before we can even start the handshake
+with the engine. Part of this is that we need to require so many files before
+we can start the work with the engine.
+
+I tried moving the setup that happens in `app/index.native` to happen before.
+But I didn't see a signifcant startup time difference. My guess is that without
+changing all files necessary to start the bootstrapping, we still load most of
+the app. So there isn't a real difference.
+
+
+### Areas to improve (in order or biggest impact and easiest to implement)
+
+* Use an alternate routing scheme that will load the screens when required or
+  when idle.
+  * this is involved, but we are doing this anyways. It would have a huge
+    impact since we wait around 2s everytime before we can do anything! That's
+    2s from the point we can start running JS to the point we've required all
+    the files we need to start the bootstrap process.
+* Minimal time to engine handshake. By requiring less files before we can start
+  the engine handshake, we can get a headstart on this and lower the time until
+  we can get useful data back from the engine.
+* Investigate if we can get by with loading a minimal version of emoij data
+
+
+## Case study: Testing immediate engine calls during bootstrap.
+
+### Abstract
+
+Normally the engine call is setup and called with a `setImmediate`. This is
+usually exactly what we want. Trading a bit of response time from the engine
+for a more response ui. During bootstrapping, our priorities our a little
+different. We can't do anything until we've bootstrapped with the engine, so it
+makes sense to avoid that tradeoff.
+
+We'll test if avoiding `setImmediate` during bootstrapping will help reduce
+startup time.
+
+### Test setup
+
+I cherry picked these commits to start the testing process:
+
+* `bd32cdee4c` - Android perf helper files
```

Comment:
Did you push these commits?
```
$ git cat-file -t bd32cdee4c
fatal: Not a valid object name bd32cdee4c
```

Reply:
(same for `507f` and `770c`)

Reply:
whoops! You're right, I just pushed it and swapped these with links